bindings/experimental/distrdf/python/DistRDF/HeadNode.py

```diff
@@ -389,4 +390,20 @@ def build_ranges(self):
                 ("Cannot build a distributed RDataFrame with zero entries. "
                  "Distributed computation will fail. "))
 
-        return Node.RangesBuilder(self)._get_clustered_ranges(self.treename, self.inputfiles, self._get_friend_info())
+        logger.debug("Building ranges for tree %s with the "
+                     "following input files:\n%s", self.treename, self.inputfiles)
+
+        # Retrieve a tuple of clusters for all files of the tree
+        clustersinfiles = Ranges.get_clusters(self.treename, self.inputfiles)
+        numclusters = len(clustersinfiles)
+
+        # Restrict `npartitions` if it's greater than clusters of the dataset
+        if self.npartitions > numclusters:
+            msg = ("Number of partitions is greater than number of clusters "
+                   "in the dataset. Using {} partition(s)".format(numclusters))
+            warnings.warn(msg, UserWarning, stacklevel=2)
```

Comment:
@eguiraud your example would not run the code here, would it? Here it's `TTreeHeadNode`, not `EmptySourceHeadNode`.

And for the TTree case this looks like a scenario where we can go on (and that's why we issue a warning). The requested number of partitions is higher than the real number of clusters, so we warn the user of that fact and we partition based on the number of clusters.

Reply:
fair enough, instead of `1` in my example put a tree with one cluster. why does that warrant a warning?

Reply:
If you have a tree with 1 cluster and ask DistRDF to create 2 partitions, DistRDF warns you that 2 > 1 and that it'll just use 1 partition, and then it goes on with running your app. What you are saying is that you'd prefer to fail if the partitioning requirements of the user cannot be met?

Reply:
I think Enrico's point is more towards the default amount of partitions being set to `2`. Also that could be changed to 1 or `None` and we could just ask the user for the partitions. But this change first needs a more robust reducer that would smartly become a no-op in case of only 1 partition

Reply:
I have added a TODO here as well

Reply:
Sorry I missed Enric's last reply, it's totally fair to warn users if _they_ specified they want N partitions, but not if they didn't specify anything